pwiz_tools/Shared/Common/Chemistry/Formula.cs

```diff
@@ -197,146 +198,230 @@ public bool Remove(string key)
-            if (parts.Length > 1)
+
+            var resultDict = new Dictionary<string, int>(this);
+            foreach (var kvpOther in other)
             {
-                var subtractive = ParseToDictionary(parts[1]);
-                foreach (var element in subtractive)
+                if (TryGetValue(kvpOther.Key, out var countCurrent))
                 {
-                    int previous;
-                    if (result.TryGetValue(element.Key, out previous))
+                    var newCount = countCurrent - kvpOther.Value;
+                    if (newCount == 0)
                     {
-                        int newCount = previous - element.Value;
-                        if (newCount == 0)
-                        {
-                            result.Remove(element.Key);
-                        }
-                        else
-                        {
-                            result[element.Key] = newCount;
-                        }
+                        resultDict.Remove(kvpOther.Key);
                     }
                     else
                     {
-                        result.Add(element.Key, -element.Value);  // Seems weird, but possibly describing a proton lost from something other that H?
+                        resultDict[kvpOther.Key] = newCount;
                     }
                 }
+                else
+                {
+                    resultDict.Add(kvpOther.Key, -kvpOther.Value);
+                }
             }
-            return result;
+
+            return FromDict(resultDict);
         }
 
-        // Subtract other's atom counts from ours
-        public T Difference(T other)
+        public T Plus(T other)
         {
+            if (other == null || other.Count == 0)
+            {
+                return (T)this;
+            }
+
             var resultDict = new Dictionary<string, int>(this);
-            foreach (var kvp in other)
+            foreach (var kvpOther in other)
             {
-                TryGetValue(kvp.Key, out int count);
-                count -= kvp.Value;
-                if (count == 0)
+                if (TryGetValue(kvpOther.Key, out var count))
                 {
-                    resultDict.Remove(kvp.Key);
+                    var newCount = count + kvpOther.Value;
+                    if (newCount == 0)
+                    {
+                        resultDict.Remove(kvpOther.Key);
+                    }
+                    else
+                    {
+                        resultDict[kvpOther.Key] = newCount;
+                    }
                 }
                 else
                 {
-                    resultDict[kvp.Key] = count;
+                    resultDict.Add(kvpOther.Key, kvpOther.Value);
                 }
             }
 
             return FromDict(resultDict);
         }
 
+        public bool IsEmpty()
+        {
+            return Dictionary == null || Dictionary.Count == 0;

```

Comment:
Is "Dictionary" really allowed to be null here? That often leads to cases where values end up slightly different when you round-trip to XML.

Reply:
Unsure - just a belt-and-suspenders thing. But I'm not sure how that affects any string representation?